app/routes.py

```diff
@@ -1 +1,246 @@
+        return {"details": "Invalid data"}, 400
+
+    if "completed_at" in request_body:
+        new_task = Task(title=request_body["title"], 
+                        description=request_body["description"],
+                        completed_at=request_body["completed_at"])
+    else:
+        new_task = Task(title=request_body["title"], 
+                description=request_body["description"])
+
+    db.session.add(new_task)
+    db.session.commit()
+
+    response = {
+        "task": new_task.create_task_dict()
+    }
+    return response, 201
+
+@tasks_bp.route("", methods=["GET"])
+def get_all_tasks():
+    title_sort_query = request.args.get("sort")
+    if title_sort_query == "asc":
+        tasks = Task.query.order_by(Task.title.asc())
+    elif title_sort_query == "desc":
+        tasks = Task.query.order_by(Task.title.desc())
+    else:
+        tasks = Task.query.all()
+
+    response = []
+    for task in tasks:
+        response.append(task.create_task_dict())
+
+    return jsonify(response)
+
+@tasks_bp.route("/<task_id>", methods=["GET"])
+def get_one_task(task_id):
+    task = get_one_task_or_abort(task_id)
+    response = {
+        "task": task.create_task_dict()
+    }
+    return response
+
+def get_one_task_or_abort(task_id):
+    try:
+        task_id = int(task_id)
+    except ValueError:
+        response = {"msg":f"Invalid task id: {task_id}. ID must be an integer."}
+        abort(make_response(jsonify(response), 400))
+    
+    requested_task = Task.query.get(task_id)
+    if requested_task is None:
+        response = {"msg":f"Could not find task with id: {task_id}"}
+        abort(make_response(jsonify(response), 404))
+        
+    return requested_task
+
+@tasks_bp.route("/<task_id>", methods=["PUT"])
+def replace_task(task_id):
+    task = get_one_task_or_abort(task_id)
+    request_body = request.get_json()
+
+    task.title = request_body["title"]
+    task.description = request_body["description"]
+
+    db.session.commit()
+    response = {
+        "task": task.create_task_dict()
+    }
+    return response
+
+@tasks_bp.route("/<task_id>", methods=["DELETE"])
+def delete_task(task_id):
+    task = get_one_task_or_abort(task_id)
+    db.session.delete(task)
+    db.session.commit()
+    response = {"details": f'Task {task.task_id} "{task.title}" successfully deleted'}
+    return response
+
+@tasks_bp.route("/<task_id>/mark_complete", methods=["PATCH"])
+def mark_task_complete(task_id):
+    task = get_one_task_or_abort(task_id)
+    task.completed_at = datetime.now()
+
+    db.session.commit()
+    response = {
+        "task": task.create_task_dict()
+    }
+
+    path = "https://slack.com/api/chat.postMessage"
+    data = {
+        "channel": "task-notifications",
+        "text": f"Someone just completed the task {task.title}"
+    }
+    headers = {
+        "authorization": "Bearer " + os.environ.get("SLACKBOT_API_KEY")
+    }
+
+    post_message = requests.post(path, data=data, headers=headers)
```

Comment:
Style: Since you don't do anything with this variable it's a best practice not to store it:

```suggestion
    requests.post(path, data=data, headers=headers)
```